src/Diagnostics/CodeAnalysis/Core/MetaAnalyzers/RegisterActionAnalyzer.cs

```diff
@@ -245,10 +371,172 @@ protected override void AnalyzeNode(SymbolAnalysisContext symbolContext, TInvoca
+                var contextParameter = model.GetSymbolInfo(receiver, cancellationToken).Symbol as IParameterSymbol;
+                if (contextParameter == null)
+                {
+                    return;
+                }
+                
+                // Check if we are bailing out on this context parameter.
+                if (_startAnalysisContextParamsToSkip != null && _startAnalysisContextParamsToSkip.Contains(contextParameter))
+                {
+                    return;
+                }
+
+                _nestedActionsMap = _nestedActionsMap ?? new Dictionary<IParameterSymbol, List<NodeAndSymbol>>();
+                List<NodeAndSymbol> registerInvocations;
+                if (!_nestedActionsMap.TryGetValue(contextParameter, out registerInvocations))
+                {
+                    registerInvocations = new List<NodeAndSymbol>();
+                }
+
+                registerInvocations.Add(new NodeAndSymbol { Invocation = invocation, Method = method });
+                _nestedActionsMap[contextParameter] = registerInvocations;
+            }
+
+            private void CodeBlockEndAction(CodeBlockAnalysisContext codeBlockContext)
+            {
+                if (_declaredStartAnalysisContextParams == null)
+                {
+                    // No declared context parameters to analyze.
+                    return;
+                }
+
+                foreach (var contextParameter in _declaredStartAnalysisContextParams)
+                {
+                    // Check if we must bail out on this context parameter.
+                    if (_startAnalysisContextParamsToSkip != null && _startAnalysisContextParamsToSkip.Contains(contextParameter))
+                    {
+                        continue;
+                    }
+
+                    var hasEndAction = false;
+                    var hasNonEndAction = false;
+
+                    List<NodeAndSymbol> registeredActions = null;
+                    if (_nestedActionsMap != null && _nestedActionsMap.TryGetValue(contextParameter, out registeredActions))
+                    {
+                        foreach (var invocationInfo in registeredActions)
+                        {
+                            switch (invocationInfo.Method.Name)
+                            {
+                                case RegisterCompilationEndActionName:
+                                case RegisterCodeBlockEndActionName:
+                                    hasEndAction = true;
+                                    break;
+
+                                default:
+                                    hasNonEndAction = true;
+                                    break;
+                            }
+                        }
+                    }
+
+                    // Report a diagnostic if no non-end actions were registered on start analysis context parameter.
+                    if (!hasNonEndAction)
```

Comment:
However, I have tried to detect cases where a context parameter is passed as argument, we bail out on analyzing such context parameter and their registrations completely. However, it can be fooled if you register a syntax node action and pass it to a separate method that registers just the end action.